### DISCUSSION TITLE
Updating FlashBuilder to use AirSDK 33 gives error during build

### DISCUSSION THREAD

**Discussion by markdemich** [airsdk/Adobe-Runtime-Support#2135] [Q&A] [answered]
I have a native air application using flex 4.1.0 and originally using version 32 of the air sdk.  However, there is a bug with the latest Mac OS so that when the user types D or e, they get some system dialog.  Apparently, that's been fixed in version 33 of the air sdk.  I download the version of 33 meant to be overlaid on flex so I did that, but now when I try to build in FlashBuilder, I see this.

Any ideas on how to fix this or debug it so I can figure out how to fix it?

Description	Resource	Path	Location	Type
1044: Interface method addPreloadedRSL in namespace mx.core:IFlexModuleFactory not implemented by class _ed33d5e0d1c1b962a6b9f6553045aca029de5b01bb5ecbea4d230163f4cfbbb2_mx_core_FlexModuleFactory.	TDXAuthoring		line 24	Flex Problem
1044: Interface method get allowDomainsInNewRSLs in namespace mx.core:IFlexModuleFactory not implemented by class _ed33d5e0d1c1b962a6b9f6553045aca029de5b01bb5ecbea4d230163f4cfbbb2_mx_core_FlexModuleFactory.	TDXAuthoring		line 24	Flex Problem
1044: Interface method get allowInsecureDomainsInNewRSLs in namespace mx.core:IFlexModuleFactory not implemented by class _ed33d5e0d1c1b962a6b9f6553045aca029de5b01bb5ecbea4d230163f4cfbbb2_mx_core_FlexModuleFactory.	TDXAuthoring		line 24	Flex Problem
1044: Interface method set allowDomainsInNewRSLs in namespace mx.core:IFlexModuleFactory not implemented by class _ed33d5e0d1c1b962a6b9f6553045aca029de5b01bb5ecbea4d230163f4cfbbb2_mx_core_FlexModuleFactory.	TDXAuthoring		line 24	Flex Problem
1044: Interface method set allowInsecureDomainsInNewRSLs in namespace mx.core:IFlexModuleFactory not implemented by class _ed33d5e0d1c1b962a6b9f6553045aca029de5b01bb5ecbea4d230163f4cfbbb2_mx_core_FlexModuleFactory.	TDXAuthoring		line 24	Flex Problem

**Comment by markdemich** [comment]
I did get some more info on this that may help someone give me some advice.  I have several component libraries, but only one that would not compile so I started comparing them.  As it turns out, for this particular one, I needed to access flash.display.NativeWindow and for that to work, I had selected "Include Adobe Air libraries" in the Flex Library Compiler options dialog in FlashBuilder 4.

I also noticed that if you select that, it will add this libraryPath 
```
    <libraryPath defaultLinkType="1">
      <libraryPathEntry kind="4" path="">
        <modifiedEntries>
          <libraryPathEntry index="6" kind="3" linkType="4" path="${PROJECT_FRAMEWORKS}/libs/rpc.swc" useDefaultLinkType="true">
            <crossDomainRsls>
              <crossDomainRslEntry autoExtract="true" policyFileUrl="http://fpdownload.adobe.com/pub/swz/crossdomain.xml" rslUrl="http://fpdownload.adobe.com/pub/swz/flex/4.1.0.16076/rpc_4.1.0.16076.swz"/>
              <crossDomainRslEntry autoExtract="true" policyFileUrl="" rslUrl="rpc_4.1.0.16076.swz"/>
            </crossDomainRsls>
          </libraryPathEntry>
          <libraryPathEntry index="5" kind="3" linkType="4" path="${PROJECT_FRAMEWORKS}/libs/sparkskins.swc" useDefaultLinkType="true">
            <crossDomainRsls>
              <crossDomainRslEntry autoExtract="true" policyFileUrl="http://fpdownload.adobe.com/pub/swz/crossdomain.xml" rslUrl="http://fpdownload.adobe.com/pub/swz/flex/4.1.0.16076/sparkskins_4.1.0.16076.swz"/>
              <crossDomainRslEntry autoExtract="true" policyFileUrl="" rslUrl="sparkskins_4.1.0.16076.swz"/>
            </crossDomainRsls>
          </libraryPathEntry>
        </modifiedEntries>
      </libraryPathEntry>
```
I'm suspecting the discrepancy has something to do with this.  I think I need a version of rpc.swc and sparkskins.swc that is built with the new air sdk?  Does that sound like it makes sense?

**Reply by markdemich** [reply]
I just moved my entire flash builder project to VS code and I'm using the latest air sdk 33.1.1.935 overlaid on Flex 4.1.  I got everything compiling and debugging.  However, when I go to make a bundle I get similar errors.

In this case they look like this.

_XXXX_mx_managers_SystemManager.as(26): col: 14 Error: Interface method addPreloadedRSL in namespace mx.core:IFlexModuleFactory not implemented by class _XXXX_mx_managers_SystemManager.

Where XXXX is the name of my application.

Any ideas on what to look for to fix this?  Do you think it's related to Flex 4.1?  I can try upgrading but I prefer not to since I'm sure that will uncover other issues.

**Reply by ajwfrost** [reply]
These are a little odd, if they worked previously with your Flex SDK. I wondered if something changed in the compiler mechanism (we get a lot of problems if the Flex compiler is overwritten with the AIR one) ... although if it was just a compiler overwrite, I would have thought it would have been apparent a little earlier.

But the issue there is because the Flex compiler generates a whole bunch of additional classes for your application, including this main start-up class based on the Flex SystemManager  one. So it seems that the mx.core.IFlexModuleFactory interface, which your SystemManager implements, has got a method (addPreloadedRSL) that the Flex compiler is not generating in your app.

Thinking about it, this may be some discrepancy between the Flex compiler version and the swc libraries that are used by Flex, so that would be something to look at. Did you update any of them? "addPreloadedRSL" was added in Flex 4.5 it seems:
https://help.adobe.com/en_US/FlashPlatform/reference/actionscript/3/mx/core/IFlexModuleFactory.html#addPreloadedRSL()
so maybe the compiler is picking up a Flex library from a later version than 4.1....

Maybe also worth checking the flex-config.xml file - and the air-config.xml file - to make sure that these are still pointing to the right places in terms of the library paths etc.

Hope that helps! Flex is very clever in what it does, but also very complex, which makes it tricky to work these sorts of things out...

**Reply by markdemich** [reply]
Thanks for this info.  It's helpful.  To be clear, I'm getting more than one of these errors.  I listed them in my original post.  

I set the dump-config flag for both flash builder and my new vs code setup and I've been comparing the config file to see of there are any suspect differences.  One thing I did notice is the root tag on the config file itself.

Under my new set up is says <royale-config> vs. <flex-config>.  Do you think that is expected?

The library-path is done differently, in flash builder it created a bunch of unique entries pointing directly to swc files, where in the new setup, it just refers to folders in the following order.
libs
libs/air

I did attempt to make them match, but that didn't fix anything.

Another difference I noticed is that in FB it sets use-resource-bundle-metadata to true whereas in the new setup it does not.

Also, in the new set up it sets exclude-native-js-library to true, but that doesn't seem to be related from what I understand.

The new setup also includes a bunch of runtime-shared-library-path entries, but maybe that's because I'm creating a "bundle" in the new setup.

The new setup has static-link-runtime-shared-libraries = true vs. false for the FB setup.

If anything pops out at you as being odd, let me know.  My plan now is to turn the flag that leave the actionscript files around that flex creates and see if I can gain any insight into what is in them and what may be amiss.

**Reply by markdemich** [reply]
I just looked at the result .as file for the generated class that's giving the errors.  Here's a snippet from it.
```
package
{

import flash.display.LoaderInfo;
import flash.text.Font;
import flash.text.TextFormat;
import flash.text.engine.TextBlock;
import flash.text.engine.TextLine;
import flash.system.ApplicationDomain;
import flash.system.Security
import flash.utils.Dictionary;
import flash.utils.getDefinitionByName;
import flashx.textLayout.compose.ISWFContext;
import mx.core.IFlexModule;
import mx.core.IFlexModuleFactory;
import mx.core.EmbeddedFontRegistry;
import mx.core.Singleton;
import mx.preloaders.DownloadProgressBar;
import mx.preloaders.SparkDownloadProgressBar;
import mx.core.FlexVersion;
import mx.styles.StyleManager;
import mx.core.mx_internal;
import mx.managers.SystemManager;

public class _XXXX_mx_managers_SystemManager
    extends mx.managers.SystemManager
    implements IFlexModuleFactory, ISWFContext
{
.....
}
```
Based on the error message it seems the IFlexModuleFactory is defining interface methods that this generated class is NOT implementing.  From what I understand, the flex compiler which is my old 4.1 compiler is building this file. Is my understanding correct?

How do I find out where mx.core.IFlexModuleFactory is defined?  Is it possible it's not using the same one that comes with Flex 4.1?

**Reply by ajwfrost** [reply]
I don't know whether you're able to use a reverse-engineer type option in an IDE, to check the definition of IFlexModuleFactory? But yes it seems like that may be the wrong version, somehow it's picking up a newer form of that interface. 

IFlexModuleFactory is defined in "core.swc" in your Flex SDK "frameworks\libs" folder. So you should be able to see where that file is being picked up, from your configuration XML file. You mention it's just "libs" as a folder but do you know what the base location for all that is? 

> One thing I did notice is the root tag on the config file itself.
> Under my new set up is says vs. . Do you think that is expected?

That seems odd... do you mean when you open the config file (flex-config.xml? or air-config.xml? from the Flex SDK "frameworks" folder?) -> it should always have `<flex-config>` as the root element.

The rest of your differences seemed okay, and I would always suggest statically linking RSLs. I'm not sure whether you can check where the libraries are being picked up from: maybe search for all instances of "core.swc" and then rename those so that the compiler fails, then you can see when it starts working again if you rename them back one by one...?

thanks

**Comment by markdemich** [comment]
I'm going to try your suggestion on figuring out which core swc file is
being used .  It may be more about the flex compiler though.  I suspect I'm
linking the correct core.swc, but the flex compiler that I'm using may be
producing the wrong code.  I was trying to figure out how to download a
fresh copy of flex 4.1 just in case mine was corrupted but I can't seem to
find anywhere to download it from.  Any ideas on that?

Also, in reference to that royal-config.  I'm using that Bowler Hat
extension with an asconfig file set to config:"air".  I had set dump-config
to output the config and the config file produced has the root element
royale-config.  I wonder if I have some misconfiguration or if there is a
bug in the extension.  However, I'm not sure that's my immediate problem
since I get this IFlexModuleFactory issue even from Flash Builder where.

You've given me a lot of insight into all this.  I'm going to dig around
some more and see what I can figure out.  Thanks again for your help.

**Reply by ajwfrost** [reply]
Hi

It may be that the Bowler Hat extension is using its own compiler perhaps? which would make sense given the `royale-config` configuration file ... but, this seems the wrong way round: I believe the compiler is responsible for generating the system manager class, and if it's missing out an interface function which was introduced in Flex 4.5 then it implies the compiler is older than the core.swc that's being used....

Getting hold of a fresh version of Flex 4.1 is going to be nigh on impossible, I think it's only 4.16 that's available now. If it's possible for you to update, and compile everything on that latest codebase, it may actually help....?

thanks

**Comment by markdemich** [comment]
Ok, I'll try to update my flex, but something I figured out.  I don't even
have core.swc.  I have a flex.sec that has that IFlexModuleFactory and if I
decompile I see the correct (lesser) number of interface methods.  For the
life of me I tried playing with the ordering of the library-path and just
about anything I can figure out and I can't tell where it thinks that
interface has more methods than what is generated by the compiler.  It's
very strange.

**Comment by markdemich** [comment, marked as answer]
@ajwfrost , I finally figured this out without needing to change my version of Flex. The culprit is a file under frameworks/libs/air named gamepad.swc. I decompiled it and it's redefining all those mx classes. It was also somehow causing my config file (via the dump-config switch) to have that root element along with a bunch of other "royal" references in the config.

To fix this I simply renamed gamepad.swc to gamepad.swc_disabled and now I'm compiling again. It's even building in Flash Builder as well.

Thanks for your help on this painful journey. Your replies did help me understand how some of this works behind the scenes. I still can't really explain though how that swc affected the dumped config file. If you have any theories, I would love to hear them.

**Reply by ajwfrost** [reply]
Wow - well found! That's an odd one .. I'll check that, since I think gamepad.swc is one of those which is just built by default as part of our release process, but it's not one we've touched (at least intentionally I think!) since taking over the codebase. Maybe we can change that SWC and avoid anyone else having this sort of problem...

Thanks for your persistence and investigations!
